athenzdomains.go

```diff
@@ -0,0 +1,183 @@
+	DomainMap       map[string]CRMap
+	lock            sync.RWMutex
+}
+
+// BuildCache - generate new athenzdomains cr cache
+func BuildCache(crIndexInformer cache.SharedIndexInformer) *Cache {
+	fmt.Println("Start building AthenzDomain map cache")
+	domainMap := make(map[string]CRMap)
+	c := &Cache{
+		CrIndexInformer: crIndexInformer,
+		DomainMap:       domainMap,
+	}
+	crIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			item, ok := obj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+			}
+			c.addObj(domainMap, item)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			newItem, ok := newObj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+			}
+			c.updateObj(domainMap, newItem)
+		},
+		DeleteFunc: func(obj interface{}) {
+			item, ok := obj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+			}
+			c.deleteObj(domainMap, item)
+		},
+	})
+	return c
+}
+
+func parseData(domainMap map[string]CRMap, domainName string, item *v1.AthenzDomain) {
+	crMap := domainMap[domainName]
+	for _, role := range item.Spec.SignedDomain.Domain.Roles {
+		crMap.RoleToPrincipals[string(role.Name)] = role.RoleMembers
+	}
+
+	for _, policy := range item.Spec.SignedDomain.Domain.Policies.Contents.Policies {
+		for _, assertion := range policy.Assertions {
+			_, ok := crMap.RoleToAssertion[assertion.Role]
+			if !ok {
+				crMap.RoleToAssertion[assertion.Role] = []SimpleAssertion{}
+			}
+			effect := assertion.Effect.String()
+			resourceRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Resource)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion resource into regex format. Error: %v", err)
+			}
+			actionRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Action)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion action into regex format. Error: %v", err)
+			}
+			simpleAssert := SimpleAssertion{
+				resource: resourceRegex,
+				action:   actionRegex,
+				effect:   effect,
+			}
+			crMap.RoleToAssertion[assertion.Role] = append(crMap.RoleToAssertion[assertion.Role], simpleAssert)
+		}
+
+	}
+}
+
+func (c *Cache) addObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := domainMap[domainName]
+	if !ok {
+		roleToPrincipals := make(map[string][]*zms.RoleMember)
+		roleToAssertion := make(map[string][]SimpleAssertion)
+		crMap := CRMap{
+			RoleToPrincipals: roleToPrincipals,
+			RoleToAssertion:  roleToAssertion,
+		}
+		domainMap[domainName] = crMap
+	}
+	parseData(domainMap, domainName, item)
+	c.lock.Unlock()
+}
+
+func (c *Cache) updateObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
```

Comment:
I think we can combine add / update together as most of the logic here is the same.